### PR TITLE
 Bar charts and Axis Type column in Plot Dock 

### PR DESCRIPTION
**Pull request by mgrojo**
This PR adds support for bar charts in the Plot Dock. A string column is used as bar labels. This is specially thought for GROUP BY queries.

A new column for displaying the axis type is added, so the plot results are predictable for the user. This column stores now the internal type enumeration.

Only columns that are correctly plotted as Y axis are selectable for that.

There are caveats, like plotting more than one Y axis, since the bars overlap. To alleviate this the additional bars are translucent, but this kind of graph is maybe not very interesting. Should the user have the option to stack the bars when that makes sense?

I'm confident of the changes, but since we plan to make soon a release, I prefer to commit this as a pull request, so it can be reviewed. I'd like opinions also about the new Axis Type column, since it is not actually needed for the implementation (maybe there should be two PR, but this is already done).

imagen

### CODE REVIEW

**Comment by justinclift** [comment]
> Should the user have the option to stack the bars when that makes sense?

Definitely.

Co-incidentally, I've been looking at doing some chart building stuff recently too (mainly box plots though), as a way to get myself back into the dbhub.io stuff.  Not using Qt for it, as the Go libraries for Qt just aren't there yet for my purposes.

**Comment by justinclift** [comment]
Just created 32 and 64 bit windows builds of this, for anyone wanting to test it out:

* https://nightlies.sqlitebrowser.org/win32-onceoffs/DB%20Browser%20for%20SQLite-bar_charts_v1-win32.exe
* https://nightlies.sqlitebrowser.org/win64-onceoffs/DB%20Browser%20for%20SQLite-bar_charts_v1-win64.exe

Maybe @struck89 @Orimay? Any interest in testing this out? :smile:

**Comment by SilvioGrosso** [comment]
Hello everyone,

Just tested this new feature on Windows 10 (64 bit):  after having installed the 64.exe version.
Everything looks fine !
Thansk a lot indeed 👍 

db-manager_group-by_plot

A little glitch also present in the previous versions for the plot feature is that the top window, the one with all the settings for the colums (X, Y, Axis type), is usually hidden when you select the Plot the first time. To make it appear you must drag down the other window with your plot. 
IMHO, this is a bit confusing  for new users.
In short, when you select the plot both windows should be visible at first.
At present, only the window for the plot itself is visible and I have to drag down its upper border in order to visualize the second window with all the options (X, Y, Axis type).

If needed I can open a new issue since:
I have encountered a bug concerning the new autocompletion feature for columns.
In short, with group by statements the columns are not recognized and you are forced to write by hand their names (or you can drag them by getting them into their schema window).
In short when you write:
count-sum-etc ( the column is not autocompleted anymore into the brackets)
At first I have thought this bug was due to the brackets () but in reality the problem occurs only when you work with group by statements.

E.g:
Select sum (   autocompletion for names of columns does not work anymore...     ) as name
from table
group by x

BTW, since we are in the Pull requests topics,  it would be extremely cool to have also the "math extension" [1]  available into the next upcoming release :-)
I am available to test it on Windows if needed.

[1] https://github.com/sqlitebrowser/sqlitebrowser/issues/1224

**Comment by mgrojo** [comment]
@justinclift Thanks for the builds. I bring now more "material".

I've committed the option for the stacked bars. When they are not stacked they are grouped together, that makes much more sense as overlapped (but that was surprisingly the default layout for QCustomPlot). The option is in the context menu of the plot. Maybe it should replace the Line type and Point shape combo-boxes for bar charts but it was easier this way. Maybe some day.

I've also added an option for displaying a plot legend for any kind of plot. Maybe some day it would be draggable with the mouse :smile:

"Stacked bars" checked and "Show legend" checked:
imagen

"Stacked bars" unchecked (grouped) and "Show legend" unchecked (defaults):
imagen

@SilvioGrosso I'm glad that my enhancement to DB4S can contribute a bit to the advance of Science :smiley:

Those problems that you mention may deserve each one a new issue. The one about the plot isn't happening in my environment. Maybe a screenshot can help discover the cause. The one about completion happens for all the functions that display a "calltip". The workaround is to click on that calltip to remove it and then the completion works again. This is how QScintilla works, I don't know if we could work around it.

**Comment by justinclift** [comment]
Wow @mgrojo, you're getting SO much stuff done.  Our next release is really going to be well named. :smile:

New builds for Windows with the new commit are currently in compiling.  When they're ready (~1/2 hour) I'll update this comment with their URLs.

> BTW, since we are in the Pull requests topics, it would be extremely cool to have also the "math extension" [1] available into the next upcoming release :-)

@SilvioGrosso Good point.  I really need to get around to making that happen. :smile:

**Comment by justinclift** [comment]
Sent a tweet about this to or Twitter account too.  Might be useful for getting feedback. :smile:

&nbsp; &nbsp; https://twitter.com/sqlitebrowser/status/955160333924085761

**Comment by mgrojo** [comment]
> Wow @mgrojo, you're getting SO much stuff done. Our next release is really going to be well named. :smile:

It's currently my pastime :smile:

**Comment by SilvioGrosso** [comment]
Hello everyone,

Just tested today's new build.
Everything is fine (on Windows 10  with the new  64.exe installer).
THANKS a lot indeed for your hard work: next version is going to be awesome 👍 

plot2

@mgrojo 
> The one about the plot isn't happening in my environment

Yep. It only occours once in a while when you install a *new* version on Windows.  
For instance, today's latest version is fine since both windows for the Plot are visible at once: no need to drag :-)

@justinclift 
> I really need to get around to making that happen.

Yep. It would be really useful for Windows users 👍

File: src/PlotDock.cpp [hunk 1]
```diff
@@ -119,16 +133,32 @@ void PlotDock::updatePlot(SqliteTableModel* model, BrowseDataTableSettings* sett
...
-                    // but store the model column index in the first 16 bit and the type
-                    // in the other 16 bits
-                    uint itemdata = 0;
-                    itemdata = i << 16;
-                    itemdata |= columntype;
-                    columnitem->setData(PlotColumnField, Qt::UserRole, itemdata);
+
+                    switch (columntype) {
+                    case QVariant::DateTime:
+                        columnitem->setText(PlotColumnType, "Date/Time");
```
Comment (MKleusberg) [inline src/PlotDock.cpp]: These should have a ```tr()``` around them as well.

File: src/PlotDock.cpp [hunk 2]
```diff
@@ -119,16 +133,32 @@ void PlotDock::updatePlot(SqliteTableModel* model, BrowseDataTableSettings* sett
...
+                        break;
+                    case QVariant::Time:
+                        columnitem->setText(PlotColumnType, "Time");
+                        break;
+                    case QVariant::Double:
+                        columnitem->setText(PlotColumnType, "Numeric");
+                        break;
+                    case QVariant::String:
+                        columnitem->setText(PlotColumnType, "Label");
+                        break;
```
Comment (MKleusberg) [inline src/PlotDock.cpp]: Can you add a ```default``` branch? Without one it's producing tons of warnings for me :wink:

File: src/PlotDock.cpp (hunk 1 above)
Comment (mgrojo) [reply src/PlotDock.cpp]: Fixed.

File: src/PlotDock.cpp (hunk 2 above)
Comment (mgrojo) [reply src/PlotDock.cpp]: Fixed. I don't get any warnings, either compiler version difference or difference in flags.

**Comment by justinclift** [comment]
Fantastic! :smile:

@chrisjlocke Feel like creating some animated gif things showing how to use the new bar charts?  They'd make for awesome instructions on the wiki. :smile:

**Comment by struck89** [comment]
My apologies for coming too late to the party.
The feature looks great. Unfortunately I don't have a use case right now, but sure I will in the coming future. Thanks!

**Pull request merged**